File: leetcode.com/medium/74. Search a 2D Matrix.js
// You are given an m x n integer matrix matrix with the following two properties:

// Each row is sorted in non-decreasing order.
// The first integer of each row is greater than the last integer of the previous row.
// Given an integer target, return true if target is in matrix or false otherwise.

// You must write a solution in O(log(m * n)) time complexity.

// Example 1:

// Input: matrix = [[1,3,5,7],[10,11,16,20],[23,30,34,60]], target = 3
// Output: true
// Example 2:

// Input: matrix = [[1,3,5,7],[10,11,16,20],[23,30,34,60]], target = 13
// Output: false

// Constraints:

// m == matrix.length
// n == matrix[i].length
// 1 <= m, n <= 100
// -104 <= matrix[i][j], target <= 104

/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
const searchMatrix = (matrix, target) => {
  const m = matrix.length
  const n = matrix[0].length

  let l = 0
  let r = m * n - 1
  while (l <= r) {
    const mid = Math.floor((l + r) / 2)
    const midV = matrix[Math.floor(mid / n)][mid % n]

    if (target === midV) return true
    if (target < midV) r = mid - 1
    if (target > midV) l = mid + 1
  }

  return false
}

File: leetcode.com/medium/74. Search a 2D Matrix.kt
// Write an efficient algorithm that searches for a value in an m x n matrix. This matrix has the following properties:

// Integers in each row are sorted from left to right.
// The first integer of each row is greater than the last integer of the previous row.
// Example 1:

// Input:
// matrix = [
//   [1,   3,  5,  7],
//   [10, 11, 16, 20],
//   [23, 30, 34, 50]
// ]
// target = 3
// Output: true
// Example 2:

// Input:
// matrix = [
//   [1,   3,  5,  7],
//   [10, 11, 16, 20],
//   [23, 30, 34, 50]
// ]
// target = 13
// Output: false

class Solution {
  fun searchMatrix(matrix: Array<IntArray>, target: Int): Boolean {
      if(matrix.size == 0 || (matrix.size == 1 && matrix[0].size == 0)) return false
      var start = 0
      var end = matrix.size - 1
      var targetArr: IntArray? = null
      while(start <= end){
          val middle = (start + end) / 2
          if(matrix[middle][matrix[middle].size - 1] < target){
              start = middle + 1
          }else if(matrix[middle][0] > target){
              end = middle - 1
          }else{
              targetArr = matrix[middle]
              break
          }
      }
      if(targetArr == null) return false
          
      start = 0
      end = targetArr.size - 1
      while(start <= end){
          val middle = (start + end) / 2
          if(targetArr[middle] < target ){
              start = middle + 1
          }else if(targetArr[middle] > target){
              end = middle - 1
          }else{
              return true
          }
      }
      return false
  }
}